pkg/identity/email/issuer.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package email
+
+import (
+	"context"
+
+	"github.com/sigstore/fulcio/pkg/identity"
+)
+
+type emailIssuer struct {
+	issuerURL string
+}
+
+func Issuer(issuerURL string) identity.Issuer {
+	return &emailIssuer{issuerURL: issuerURL}
+}
+
+func (e *emailIssuer) Authenticate(ctx context.Context, token string) (identity.Principal, error) {
```

Comment:
All of these implementations of Authenticate and Match are going to look the same, so I wonder if we can either implement this with generics, or implement this on the base Principal type. If the base Principal type included a `PrincipalFromIDToken` function, it could then implement the Issuer interface too.

Reply:
I don't think it makes sense to include `PrincipalFromIDToken` in the interface since it's expected that `Authenticate` will always call it, and once this refactor is done we won't need to call `PrincipalFromIDToken ` outside of a specific package. I think we can make it a private helper function once this refactor is done though, since it shouldn't need to be called outside of the package at that point!

Reply:
Let's see if there's some way to reduce common implementations, but I'm fine if it's in a later PR. Maybe just have Match be implemented on the base type? And Authenticate ends up calling a private `principcalFromIDToken`?